src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs

```diff
@@ -131,17 +131,29 @@ private void WriteHeaders(ref HeaderEncodingState state, HttpHeaders headers)
                     }
 
                     Debug.Assert(header.Value.Length > 0, "No values for header??");
+
+                    // Write cookies from CookieContainer first.
+                    if (cookiesFromContainer != null && header.Key.KnownHeader == KnownHeaders.Cookie)
```

Comment:
Why is this code inside the 'for' loop?  Presumably I think you're trying to write the 'Cookie' header from the container-based cookies next to the manual 'Cookie' headers (if present).  I don't think that's necessary. And it will slow the loop by having to check for the "if (ccokieFromContainer != null)" all the time.

It would be faster to just write out this 'Cookie' header first before entering the loop.  In fact, by doing that, you don't need to modify this method at all.  Just call one of the overloads of WriteHeader before actually calling this method which loops thru all the rest of the headers.

Reply:
> Why is this code inside the 'for' loop? Presumably I think you're trying to write the 'Cookie' header from the container-based cookies next to the manual 'Cookie' headers (if present).

Yes. I want to make all Cookie header fields are grouped together.

> I don't think that's necessary.

I think you are right. Since 1. I'm not able to find in the RFC that grouping is required, 2. move this out of the "for" loop can improve performance. Will change in next commit.